zaza/openstack/charm_tests/octavia/tests.py

```diff
@@ -236,13 +236,21 @@ def wait_for_lb_resource(octavia_show_func, resource_id,
         """Wait for loadbalancer resource to reach expected status."""
         provisioning_status = provisioning_status or 'ACTIVE'
         resp = octavia_show_func(resource_id)
-        logging.info(resp['provisioning_status'])
-        assert resp['provisioning_status'] == provisioning_status, (
-            'load balancer resource has not reached '
-            'expected provisioning status: {}'
-            .format(resp))
+        logging.info("Current provisioning status: {}, waiting for {}"
+                     .format(resp['provisioning_status'], provisioning_status))
+
+        msg = ('load balancer resource has not reached '
+               'expected provisioning status: {}'.format(resp))
+
+        # ERROR is a final state, once it's reached there is no reason to keep
+        # retrying and delaying the failure.
+        if resp['provisioning_status'] == 'ERROR':
+            raise ValueError(msg)
```

Comment:
I think raising an AssertionError would probably be a bit better than ValueError. It's effectively the same end result, but when reading the trace backs it would show that the assertion failed which is probably more practical for debugging from logs in test runs.

Reply:
I wondered about that too, @wolson, but oddly (?) the code uses tenacity and the `AssertionError` triggers a retry.  So `ValueError` is one way to abort and escape.

I think a proper fix would be to use a custom `Exception` for the 'normal' retries (i.e.  not re-purpose `AssertionError`) and then use `AssertionError` here to allow the escape.  e.g. rewrite the tenacity block with a Custom Exception, and retrieve asserts to be asserts rather that causing a retry.

Reply:
yes, I didn't go for `AssertionError` because it doesn't break the retries. I will add the custom Exception as @ajkavanagh suggests.